src/metric/survival_metric.cu

```diff
@@ -0,0 +1,304 @@
+        static_cast<double>(h_preds[i])) * wt;
+      weights_sum += wt;
+    }
+    PackedReduceResult res{residue_sum, weights_sum};
+    return res;
+  }
+
+#if defined(XGBOOST_USE_CUDA)
+
+  PackedReduceResult DeviceReduceMetrics(
+      const HostDeviceVector<bst_float>& weights,
+      const HostDeviceVector<bst_float>& labels_lower_bound,
+      const HostDeviceVector<bst_float>& labels_upper_bound,
+      const HostDeviceVector<bst_float>& preds) {
+    size_t ndata = labels_lower_bound.Size();
+    CHECK_EQ(ndata, labels_upper_bound.Size());
+
+    thrust::counting_iterator<size_t> begin(0);
+    thrust::counting_iterator<size_t> end = begin + ndata;
+
+    auto s_label_lower_bound = labels_lower_bound.DeviceSpan();
+    auto s_label_upper_bound = labels_upper_bound.DeviceSpan();
+    auto s_preds = preds.DeviceSpan();
+    auto s_weights = weights.DeviceSpan();
+
+    const bool is_null_weight = (weights.Size() == 0);
+
+    auto d_policy = policy_;
+
+    dh::XGBCachingDeviceAllocator<char> alloc;
+    PackedReduceResult result = thrust::transform_reduce(
+      thrust::cuda::par(alloc),
+      begin, end,
+      [=] XGBOOST_DEVICE(size_t idx) {
+        double weight = is_null_weight ? 1.0 : static_cast<double>(s_weights[idx]);
+        double residue = d_policy.EvalRow(
+            static_cast<double>(s_label_lower_bound[idx]),
+            static_cast<double>(s_label_upper_bound[idx]),
+            static_cast<double>(s_preds[idx]));
+        residue *= weight;
+        return PackedReduceResult{residue, weight};
+      },
+      PackedReduceResult(),
+      thrust::plus<PackedReduceResult>());
+
+    return result;
+  }
+
+#endif  // XGBOOST_USE_CUDA
+
+  PackedReduceResult Reduce(
+      int device,
+      const HostDeviceVector<bst_float>& weights,
+      const HostDeviceVector<bst_float>& labels_lower_bound,
+      const HostDeviceVector<bst_float>& labels_upper_bound,
+      const HostDeviceVector<bst_float>& preds) {
+    PackedReduceResult result;
+
+    if (device < 0) {
+      result = CpuReduceMetrics(weights, labels_lower_bound, labels_upper_bound, preds);
+    }
+#if defined(XGBOOST_USE_CUDA)
+    else {  // NOLINT
+      device_ = device;
+      preds.SetDevice(device_);
+      labels_lower_bound.SetDevice(device_);
+      labels_upper_bound.SetDevice(device_);
+      weights.SetDevice(device_);
+
+      dh::safe_cuda(cudaSetDevice(device_));
+      result = DeviceReduceMetrics(weights, labels_lower_bound, labels_upper_bound, preds);
+    }
+#endif  // defined(XGBOOST_USE_CUDA)
+    return result;
+  }
+
+ private:
+  EvalRow policy_;
+#if defined(XGBOOST_USE_CUDA)
+  int device_{-1};
+#endif  // defined(XGBOOST_USE_CUDA)
+};
+
+struct EvalIntervalRegressionAccuracy {
+  void Configure(const Args& args) {}
+
+  const char* Name() const {
+    return "interval-regression-accuracy";
+  }
+
+  XGBOOST_DEVICE double EvalRow(
+      double label_lower_bound, double label_upper_bound, double log_pred) const {
```

Comment:
Maybe just use float to remove static cast in following classes?

Reply:
We are dealing with probability distributions with log and exp, so the extra precision matters. See Appendix C of the paper https://arxiv.org/abs/2006.04920.

Reply:
@hcho3 Got it.  Exponential distributions are quite difficult to handle in practice.